File: src/minimumGLTriangle.nim
import winlean4k, openGL4k2

{.pragma: forceInline, codegenDecl: "__forceinline  $# $#$#".}

const
  ScreenWidth   = 640
  ScreenHeight  = 480

let
  pfd           = PIXELFORMATDESCRIPTOR(
    nSize:          sizeof(PIXELFORMATDESCRIPTOR).uint16,
    nVersion:       1'u16,
    dwFlags:        PFD_DRAW_TO_WINDOW or
                    PFD_SUPPORT_OPENGL or
                    PFD_DOUBLEBUFFER,
    iPixelType:     PFD_TYPE_RGBA,
    cColorBits:     24,
    cRedBits:       0, cRedShift:   0,
    cGreenBits:     0, cGreenShift: 0,
    cBlueBits:      0, cBlueShift:  0,
    cAlphaBits:     0, cAlphaShift: 0,
    cAccumBits:     0,
    cAccumRedBits:  0,
    cAccumGreenBits:0,
    cAccumBlueBits: 0,
    cAccumAlphaBits:0,
    cDepthBits:     24,
    cStencilBits:   0,
    cAuxBuffers:    0,
    iLayerType:     PFD_MAIN_PLANE,
    bReserved:      0,
    dwLayerMask:    0,
    dwVisibleMask:  0,
    dwDamageMask:   0)

proc initScreen(): auto {.inline.} =
  let hWnd = CreateWindowA(
    "STATIC".cstring, nil,
    WS_POPUP or WS_VISIBLE, 0, 0,
    ScreenWidth, ScreenHeight, nil, nil, nil, nil)
  let hdc = GetDC(hWnd)

  discard SetPixelFormat(hdc, ChoosePixelFormat(hdc, unsafeAddr pfd), unsafeAddr pfd)
  discard wglMakeCurrent(hdc, wglCreateContext(hdc))
  loadExtensions()

  ShowCursor(0)

  return hdc

proc createShaderProgram(source:cstring, shaderType: GLEnum): GLuint {.inline.}=
  result = glCreateShaderProgramv(shaderType, 1, cast[cstringArray](unsafeAddr source))
  assert result != 0
  when not defined(danger):
    var logLen: GLint
    glGetProgramiv(result, GL_INFO_LOG_LENGTH, addr logLen)
    if logLen != 0:
      var log = newseq[int8](logLen)
      glGetProgramInfoLog(result, logLen, nil, cast[cstring](addr log[0]))
      echo "Message from OpenGL shader compiler:"
      echo cast[cstring](addr log[0])

    var success: GLint
    glGetProgramiv(result, GL_LINK_STATUS, addr success)
    if success != cast[GLint](GL_TRUE):
      quit "Failed to link shader"

const triangleVSSrc = staticRead("../shaders/triangle.vs").cstring
const triangleFSSrc = staticRead("../shaders/triangle.fs").cstring

proc initScene() {.forceInline.} =
  let vpo = createShaderProgram(triangleVSSrc, GL_VERTEX_SHADER)
  let fpo = createShaderProgram(triangleFSSrc, GL_FRAGMENT_SHADER)
  var pipeObj: GLuint
  glCreateProgramPipelines(1, addr pipeObj)
  glUseProgramStages(pipeObj, GL_VERTEX_SHADER_BIT, vpo)
  glUseProgramStages(pipeObj, GL_FRAGMENT_SHADER_BIT, fpo)
  glBindProgramPipeline(pipeObj)

proc WinMainCRTStartup() {.exportc.} =
  let hdc = initScreen()
  initScene()

  discard wglSwapIntervalEXT(1)

  var msg{.noinit.}: MSG

  while true:
    discard PeekMessageA(addr msg, nil, 0, 0, PM_REMOVE)
    glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)
    glDrawArrays(GL_TRIANGLES, 0, 3)
    discard SwapBuffers(hdc)
    if GetAsyncKeyState(VK_ESCAPE) != 0:
      ExitProcess(0)

  # Process keep alive if ExitProcess API was not called.
  ExitProcess(0)

when not defined(danger):
  WinMainCRTStartup()

File: csrc/minimumGLTriangle.c
/*
 * Minimum C program.
 * Nim program can be as small as this program?
 * Run minimum_build.bat on VS2015 x86 Native Tools command prompt to build this code.
 */

#include <windows.h>
#include <GL/gl.h>
#include <GL/glext.h>
#include <GL/wglext.h>

#define WIDTH 640
#define HEIGHT 480

PFNGLCREATESHADERPROC glCreateShader;
PFNGLSHADERSOURCEPROC glShaderSource;
PFNGLCOMPILESHADERPROC glCompileShader;
PFNGLCREATEPROGRAMPROC glCreateProgram;
PFNGLLINKPROGRAMPROC glLinkProgram;
PFNGLATTACHSHADERPROC glAttachShader;
PFNGLUSEPROGRAMPROC glUseProgram;
PFNWGLSWAPINTERVALEXTPROC wglSwapIntervalEXT;

__forceinline void loadExtensions() {
  glCreateShader = (PFNGLCREATESHADERPROC)wglGetProcAddress("glCreateShader");
  glShaderSource = (PFNGLSHADERSOURCEPROC)wglGetProcAddress("glShaderSource");
  glCompileShader = (PFNGLCOMPILESHADERPROC)wglGetProcAddress("glCompileShader");
  glCreateProgram = (PFNGLCREATEPROGRAMPROC)wglGetProcAddress("glCreateProgram");
  glLinkProgram = (PFNGLLINKPROGRAMPROC)wglGetProcAddress("glLinkProgram");
  glAttachShader = (PFNGLATTACHSHADERPROC)wglGetProcAddress("glAttachShader");
  glUseProgram = (PFNGLUSEPROGRAMPROC)wglGetProcAddress("glUseProgram");
  wglSwapIntervalEXT = (PFNWGLSWAPINTERVALEXTPROC)wglGetProcAddress("wglSwapIntervalEXT");
}

static PIXELFORMATDESCRIPTOR pfd =
{
	sizeof(PIXELFORMATDESCRIPTOR),
	1,									          //OpenGL version
	PFD_DRAW_TO_WINDOW |
	PFD_SUPPORT_OPENGL |
	PFD_DOUBLEBUFFER,
	PFD_TYPE_RGBA,
	24,									//Color bits
	0, 0,
	0, 0,
	0, 0,
	0, 0,
	0,
	0, 0, 0, 0,
	24,									//Z buffer
	0,
	0,
	PFD_MAIN_PLANE,
	0,
	0, 0, 0
};

__forceinline HDC initScreen() {
	HWND hWnd =	CreateWindowA("STATIC", 0,
		WS_POPUP|WS_VISIBLE, 0, 0,
		WIDTH, HEIGHT, NULL, NULL, NULL,NULL);

	HDC hdc = GetDC(hWnd);

	SetPixelFormat(hdc, ChoosePixelFormat(hdc, &pfd), &pfd);
	wglMakeCurrent(hdc, wglCreateContext(hdc));

  loadExtensions();

  return hdc;
}

__forceinline GLuint createShader(const char* source, GLenum shaderType) {
  GLenum result = glCreateShader(shaderType);
  glShaderSource(result, 1, (const GLchar**)&source, NULL);
  glCompileShader(result);
  
  return result;
}

__forceinline void linkProgramObj(GLuint progObj) {
  glLinkProgram(progObj);
}

#include "../shaders/triangle.vs.h"
#include "../shaders/triangle.fs.h"

__forceinline void initScene() {
  GLuint vso = createShader(triangle_vs, GL_VERTEX_SHADER);
  GLuint fso = createShader(triangle_fs, GL_FRAGMENT_SHADER);
  GLuint progObj = glCreateProgram();
  glAttachShader(progObj, vso);
  glAttachShader(progObj, fso);
  linkProgramObj(progObj);
  glUseProgram(progObj);
}

void WinMainCRTStartup(void) {
  HDC hdc = initScreen();
  initScene();

  wglSwapIntervalEXT(1);

	MSG msg;
loop:
	PeekMessage(&msg, 0, 0, 0, PM_REMOVE);
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
  glDrawArrays(GL_TRIANGLES, 0, 3);
	SwapBuffers(hdc);
	if(GetAsyncKeyState(VK_ESCAPE))
		ExitProcess(0);
	goto loop;
}
